logdetective/server/metric.py

```diff
@@ -0,0 +1,62 @@
+import datetime
+import inspect
+from typing import Union
+from functools import wraps
+
+from starlette.responses import StreamingResponse
+from logdetective.server import database, models
+
+
+def track_request():
+    """
+    Decorator to track requests metrics
+    """
+
+    def add_new_metrics(api_name: str, build_log: models.BuildLog) -> int:
+        return database.model.AnalyzeRequestMetrics.create(
+            endpoint=database.model.AnalyzeRequestMetrics.EndpointType(api_name),
+            log_url=build_log.url,
+            request_received_at=datetime.datetime.now(datetime.timezone.utc),
+        )
+
+    def update_metrics(
+        metrics_id: int,
+        response: Union[models.Response, models.StagedResponse, StreamingResponse],
+    ) -> None:
+        response_sent_at = datetime.datetime.now(datetime.timezone.utc)
+        response_length = None
+        if hasattr(response, "explanation") and "choices" in response.explanation:
+            response_length = sum(
+                len(choice["text"])
+                for choice in response.explanation["choices"]
+                if "text" in choice
+            )
+        response_certainty = (
+            response.response_certainty
+            if hasattr(response, "response_certainty")
+            else None
+        )
+        database.model.AnalyzeRequestMetrics.update(
+            metrics_id, response_sent_at, response_length, response_certainty
+        )
```

Comment:
I think the `add_new_metrics` and `update_metrics` functions can be top-level and potentially called without the decorator if needed?

Reply:
Yes, I agree. I will move them up.